packages/amplify-graphql-auth-transformer/src/graphql-auth-transformer.ts

```diff
@@ -794,6 +799,97 @@ Static group authorization should perform as expected.`,
     );
   };
 
+  addAutoGeneratedFields = (
+    ctx: TransformerContextProvider,
+    def: ObjectTypeDefinitionNode,
+    allowedFields: string[],
+    fields: readonly string[],
+  ) => {
+    this.addAutoGeneratedRelationalFields(ctx, def, allowedFields, fields);
+    this.addAutoGeneratedIndexFields(def, allowedFields);
+  };
+
+  addAutoGeneratedRelationalFields = (
+    ctx: TransformerContextProvider,
+    def: ObjectTypeDefinitionNode,
+    allowedFields: string[],
+    fields: readonly string[],
+  ) => {
+    this.addAutoGeneratedHasManyFields(ctx, def, allowedFields);
+    this.addAutoGeneratedHasOneFields(ctx, fields, def, allowedFields);
+  };
+
+  addAutoGeneratedIndexFields = (def: ObjectTypeDefinitionNode, allowedFields: string[]) => {
+    const sortKeyFieldValues: ListValueNode[] = def.fields
+      ?.map(it => it.directives)
+      .flat()
+      .filter(it => it.name.value === 'primaryKey' || it.name.value === 'index')
+      .map(it => it.arguments)
+      .flat()
+      .filter(it => it.name.value === 'sortKeyFields' && it.value.kind === 'ListValue' && it.value.values.length > 1)
+      .map(it => it.value)
+      .flat() as ListValueNode[];
+
+    for (const sortKeyFieldValue of sortKeyFieldValues) {
+      const accessOnAllKeys = !sortKeyFieldValue.values.some(it => it.kind !== 'StringValue' || !allowedFields.includes(it.value));
+      if (!accessOnAllKeys) continue;
+
+      const keyName = sortKeyFieldValue.values.map(it => (it as StringValueNode).value).join(ModelResourceIDs.ModelCompositeKeySeparator());
+      if (!allowedFields.includes(keyName)) {
+        allowedFields.push(keyName);
+      }
+    }
+  };
+
+  addAutoGeneratedHasManyFields = (ctx: TransformerContextProvider, def: ObjectTypeDefinitionNode, allowedFields: string[]) => {
+    const typeDefs = ctx.inputDocument.definitions.filter(it => it.kind === 'ObjectTypeDefinition') as ObjectTypeDefinitionNode[];
+
+    const hasManyRelatedFields = typeDefs
+      .map(it =>
+        it.fields.map(field => {
+          return { ...field, relatedType: it.name.value };
+        }),
+      )
+      .flat()
+      .filter(it => getBaseType(it.type) === def.name.value && it.directives?.some(d => d.name.value === 'hasMany'));
+
+    for (const relatedField of hasManyRelatedFields) {
+      if (!allowedFields.includes(getConnectionAttributeName(relatedField.relatedType, relatedField.name.value))) {
```

Comment:
we would it be better to create a set and pass it down for both relational functions?

Reply:
updated